control/controlclient/direct.go

```diff
@@ -1204,6 +1221,32 @@ func sleepAsRequested(ctx context.Context, logf logger.Logf, timeoutReset chan<-
 	}
 }
 
+// getNoiseClient returns the noise client, creating one if one doesn't exist.
+func (c *Direct) getNoiseClient() (*noiseClient, error) {
+	c.mu.Lock()
+	serverNoiseKey := c.serverNoiseKey
+	nc := c.noiseClient
+	c.mu.Unlock()
+	if serverNoiseKey.IsZero() {
+		return nil, errors.New("zero serverNoiseKey")
+	}
+	if nc != nil {
+		return nc, nil
+	}
+	np, err, _ := c.sfGroup.Do("noise", func() (interface{}, error) {
+		k, err := c.getMachinePrivKey()
+		if err != nil {
+			return nil, err
+		}
+
+		return newNoiseClient(k, serverNoiseKey, c.serverURL)
+	})
+	c.mu.Lock()
+	c.noiseClient = np.(*noiseClient)
```

Comment:
this will panic if nip is nil on error.

You want `c.noiseClient, _ = ....`

Reply:
But also this part can have multiple goroutines racing on it.

You want to do the assignment inside the `Do` func.

Reply:
I'd rather just handle the error explicitly. Done

Reply:
you still want to do the assignment inside the Do

Reply:
Moved to `Do`